eth/downloader/whitelist/service.go

```diff
@@ -145,7 +164,128 @@ func (s *Service) IsValidChain(currentHeader *types.Header, chain []*types.Heade
 		return milestoneBool, err
 	}
 
-	return true, nil
+	// At last, check for fork correctness
+	return s.checkForkCorrectness(chain), nil
+}
+
+// checkForkCorrectness checks if the chain being imported belongs to the correct fork
+// by iterating backwards until the last whitelisted milestone or checkpoint. This helps
+// in cases where the chain being imported doesn't overlap with any whitelisted entry and
+// we have to blindly accept it.
+func (s *Service) checkForkCorrectness(chain []*types.Header) bool {
+	if len(chain) == 0 {
+		return true
+	}
+	firstBlock := chain[0]
+	headerNumber := firstBlock.Number.Uint64()
+	lastHeaderNumber := chain[len(chain)-1].Number.Uint64()
+
+	// Recent most whitelisted entry
+	var (
+		number uint64
+		hash   common.Hash
+	)
+
+	// Fetch the latest whitelisted entry using both checkpoint and milestone
+	milestoneExists, milestoneNumber, milestoneHash := s.milestoneService.Get()
+	checkpointExists, checkpointNumber, checkpointHash := s.checkpointService.Get()
+	if !milestoneExists && !checkpointExists {
+		return true
+	}
+	if milestoneExists {
+		number = milestoneNumber
+		hash = milestoneHash
+	}
+	if checkpointExists {
+		// Only choose checkpoint if it's more recent than last milestone
+		if checkpointNumber > number {
+			number = checkpointNumber
+			hash = checkpointHash
+		}
+	}
+
+	var lastKnownValidBlock uint64 = number
+	if s.lastValidForkBlock > number {
+		lastKnownValidBlock = s.lastValidForkBlock
+	}
+
+	// Blind accept the chain if we've to iterate more than `maxForkCorrectnessLimit` blocks
+	if headerNumber-lastKnownValidBlock > s.maxForkCorrectnessLimit {
+		log.Debug("Skipping fork correctness check as block is too far ahead", "block", headerNumber, "last whitelisted", number)
+		return true
+	}
+
+	// Track all blocks iterated for caching
+	var blocksChecked []common.Hash = make([]common.Hash, 0, DefaultMaxForkCorrectnessLimit)
+	// Cache the incoming chain by default
+	for _, header := range chain {
+		blocksChecked = append(blocksChecked, header.Hash())
+	}
+
+	// Acquire lock over the cache
+	s.forkValidationCacheMu.Lock()
+	defer s.forkValidationCacheMu.Unlock()
+
+	// Only perform checks if first block being imported is ahead of last whitelisted entry. We can
+	// skip checking for fork correctness otherwise as chain would be already validated before.
+	if headerNumber > number {
+		parentNumber := headerNumber - 1
+		parentHash := firstBlock.ParentHash
+		for {
```

Comment:
Maybe it would be safer to put a loop condition here, e.g. check no more than X number of block?

Reply:
I am already checking for how long we have to go back in past to validate the chain [here](https://github.com/0xPolygon/bor/pull/1764/files#diff-7b1971c3789e15f3b09e59dfa774e88a6ac4641fdb75d638fabd6e38d419cb74R213). If it's too long, we return directly.